index/field/multi_field_intersect_iterator.go

```diff
@@ -2,91 +2,56 @@ package field
 
 import (
 	"github.com/xichen2020/eventdb/document/field"
+	"github.com/xichen2020/eventdb/index"
 )
 
 // MultiFieldIntersectIterator is an iterator that iterates over multiple fields,
-// which are joined on their doc IDs. As a result, a document needs to
+// which are intersected on their doc IDs. As a result, a document needs to
 // contain all fields associated with the iterator in order for it to
 // be included in the output of the multi-field iterator.
 type MultiFieldIntersectIterator struct {
-	iters []BaseFieldIterator
+	iters         []BaseFieldIterator          // Used to retrieve the underlying field values
+	intersectIter *index.InAllDocIDSetIterator // Used to move the base iterators in lock step
 
-	done       bool
-	err        error
-	currDocIDs []int32
-	currVals   []field.ValueUnion
+	currVals []field.ValueUnion
 }
 
 // NewMultiFieldIntersectIterator creates a new multi-field intersecting iterator.
 func NewMultiFieldIntersectIterator(iters []BaseFieldIterator) *MultiFieldIntersectIterator {
+	docIDIters := make([]index.DocIDSetIterator, len(iters))
+	for i := 0; i < len(iters); i++ {
+		docIDIters[i] = iters[i]
+	}
 	return &MultiFieldIntersectIterator{
-		iters:      iters,
-		done:       len(iters) == 0,
-		currDocIDs: make([]int32, len(iters)),
-		currVals:   make([]field.ValueUnion, len(iters)),
+		iters:         iters,
+		intersectIter: index.NewInAllDocIDSetIterator(docIDIters...),
+		currVals:      make([]field.ValueUnion, len(iters)),
 	}
 }
 
 // Next returns true if there are more items to be iterated over.
-func (it *MultiFieldIntersectIterator) Next() bool {
-	if it.done || it.err != nil {
-		return false
-	}
-
-	// Advance all iterators first.
-	for i, iit := range it.iters {
-		if !iit.Next() {
-			it.done = true
-			it.err = iit.Err()
-			return false
-		}
-		it.currDocIDs[i] = iit.DocID()
-	}
-
-	for {
-		var minIdx, maxIdx int
-		for i := 1; i < len(it.currDocIDs); i++ {
-			if it.currDocIDs[i] < it.currDocIDs[minIdx] {
-				minIdx = i
-			}
-			if it.currDocIDs[i] > it.currDocIDs[maxIdx] {
-				maxIdx = i
-			}
-		}
-		// All iterators have the same doc ID.
-		if it.currDocIDs[minIdx] == it.currDocIDs[maxIdx] {
-			for i := 0; i < len(it.iters); i++ {
-				it.currVals[i] = it.iters[i].ValueUnion()
-			}
-			return true
-		}
-		if !it.iters[minIdx].Next() {
-			it.done = true
-			it.err = it.iters[minIdx].Err()
-			return false
-		}
-		it.currDocIDs[minIdx] = it.iters[minIdx].DocID()
-	}
-}
+func (it *MultiFieldIntersectIterator) Next() bool { return it.intersectIter.Next() }
```

Comment:
Looking at the implementation of `InAnyDocIDSetIterator`, it looks like it does not perform the intersect operation. As in, it only advances each underlying base iterator forward w/o ensuring that all DocIDs line up.

It looks like it's used below in the `MultiFieldUnionIterator` which makes sense. But in that case, aren't we missing the intersect operation here? Maybe I'm missing something though.

Reply:
Note that this uses `InAllDocIDSetIterator` which advances all doc ID iterators in sync.

Reply:
Oh woops disregard my comment then.